### PR TITLE
BUG: iloc fails with non lex-sorted MultiIndex #13797

### LINKED ISSUE #13797: BUG: iloc fails with non lex-sorted MultiIndex
#### Code Sample, a copy-pastable example if possible
```python
import pandas as pd
import numpy as np
ind = [
        ['AA','AA','AA','BB','BB'],
        ['A' ,'B' ,'B' ,'a' ,'b']
] 
ind_nonLex = [
        ['CC','CC','CC','BB','BB'],
        ['A' ,'B' ,'B' ,'a' ,'b']
] 

strCol=pd.DataFrame([['fooA'],['fooB'],['fooC'],['fooD'],['fooE']])

dat=np.arange(1,26).reshape(5,5)
df=pd.concat([strCol, pd.DataFrame(dat)], axis=1)
df1=pd.DataFrame(df.values, index=ind_nonLex)
df2=pd.DataFrame(df.values, index=ind)
```

df1 (whose index is not lex sorted) fails with iloc access:
```
>>> df1.iloc[0,0]
C:\Users\rikuhiro\Anaconda3\envs\pd-check\lib\site-packages\ipykernel\__main__.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  if __name__ == '__main__':
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-37-884ae2904642> in <module>()
----> 1 df1.iloc[0,0]

C:\Users\rikuhiro\Anaconda3\envs\pd-check\lib\site-packages\pandas\core\indexing.py in __getitem__(self, key)
   1292 
   1293         if type(key) is tuple:
-> 1294             return self._getitem_tuple(key)
   1295         else:
   1296             return self._getitem_axis(key, axis=0)

C:\Users\rikuhiro\Anaconda3\envs\pd-check\lib\site-packages\pandas\core\indexing.py in _getitem_tuple(self, tup)
   1561 
   1562             # if the dim was reduced, then pass a lower-dim the next time
-> 1563             if retval.ndim < self.ndim:
   1564                 axis -= 1
   1565 

AttributeError: 'str' object has no attribute 'ndim'
```

#### Expected Output
df2 (whose index is lex sorted) works as expected:
```python
>>> df2.iloc[0,0]
'fooA'
```

This attributeError does not occur when the DataFrame.values consist of numpy objects (e.g. numpy.int32) because they have the ndim attribute. (Although the performance warning remains, it may be another issue).

I found that the addition of an __if__ statement can remedy this in pandas/core/indexing.py. This just makes the ___getitem_tuple(self, tup)__ be aware of objects without the ndim attribute, as ___getitem_nested_tuple(self, tup)__ is (I will prepare a pull request if it is helpful.)

```python
 @@ -1569,6 +1569,10 @@ def _getitem_tuple(self, tup):
              retval = getattr(retval, self.name)._getitem_axis(key, axis=axis)
  
 +            # if we have a scalar, we are done
 +            if lib.isscalar(retval) or not hasattr(retval, 'ndim'):
 +                break
 +
              # if the dim was reduced, then pass a lower-dim the next time
              if retval.ndim < self.ndim:
                  axis -= 1
```

#### output of ``pd.show_versions()``
```
>>> pd.show_versions()
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 58 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.1
nose: None
pip: 8.1.2
setuptools: 23.0.0
Cython: None
numpy: 1.11.1
scipy: None
statsmodels: None
xarray: None
IPython: 5.0.0
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None
```

### PR DESCRIPTION
**Pull request by ygriku**
- [x] closes #13797
- [x] tests added / passed pandas/tests/test_multilevel.py
- [x] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

corrected how iloc handles tuple-keys for multiindex

### CODE REVIEW

**Comment by shoyer** [comment]
Yikes! This is a testament to how convoluted the indexing code currently is. If we did this right, it would not be possible for code using iloc to get anywhere near checks for multi-indexes.

**Comment by jorisvandenbossche** [comment]
> If we did this right, it would not be possible for code using iloc to get anywhere near checks for multi-indexes. 

But should it? For `iloc` it shouldn't matter if you have a multi-index or not how the indexer is interpreted

File: pandas/tests/test_multilevel.py [hunk 1]
```diff
@@ -2420,6 +2420,25 @@ def test_repeat(self):
...
+            ['A', 'B', 'B', 'a', 'b']
+        ]
+
+        strCol = pd.DataFrame(
+            ['fooA', 'fooB', 'fooC', 'fooD', 'fooE'])
+
+        dat = np.arange(1, 26).reshape(5, 5)
+        df = pd.concat([strCol, pd.DataFrame(dat)], axis=1)
+        df1 = pd.DataFrame(df.values, index=ind_nonLex)
+
```
Comment (jreback) [inline pandas/tests/test_multilevel.py]: construct the expected result and use ``tm.assert_frame_equal``

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -903,7 +903,7 @@ def _getitem_lowerdim(self, tup):
 
         # we maybe be using a tuple to represent multiple dimensions here
         ax0 = self.obj._get_axis(0)
-        if isinstance(ax0, MultiIndex):
+        if isinstance(ax0, MultiIndex) and self.name != 'iloc':
```
Comment (jreback) [inline pandas/core/indexing.py]: add a comment here (or can update the previous one), with the issue number and a 1-line about what is happening here

File: pandas/tests/test_multilevel.py [hunk 3]
```diff
@@ -2420,6 +2420,25 @@ def test_repeat(self):
         m_df = pd.Series(data, index=m_idx)
         assert m_df.repeat(3).shape == (3 * len(data), )
 
+    def test_iloc_mi(self):
+        # GH 13797
```
Comment (jreback) [inline pandas/tests/test_multilevel.py]: add an informative comment on what you are testing here

**Comment by jreback** [comment]
> Yikes! This is a testament to how convoluted the indexing code currently is. If we did this right, it would not be possible for code using iloc to get anywhere near checks for multi-indexes.

This is not a helpful comment. This is just an implemention detail of the indexer. It could be a sub-class but its not, otherwise this would just be an overriden method.

**Comment by shoyer** [comment]
> This is not a helpful comment. This is just an implemention detail of the indexer. It could be a sub-class but its not, otherwise this would just be an overriden method.

This fix looks fragile to me. That said, it's probably the right choice for now. It fixes the problem at hand, but its not obvious that it's fully correct, because too many different code paths reuse the same logic in different ways.

**Comment by jreback** [comment]
can you rebase / update?

**Comment by YG-Riku** [comment]
Yes. I can probably find a time tomorrow.

**Comment by YG-Riku** [comment]
@jreback 
Sorry for taking this long time. I added some comments, and modified the test code so that the test purpose became clear (i.e. iloc handles tuple as an integer locator instead of trying to interpret as multiindex representation.)

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**